### PR TITLE
Adding SIKE

### PR DESCRIPTION
**Pull request by drucker-nir**
**Issue # (if available):** 

**Description of changes:** 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pq-crypto/bike/defs.h [hunk 1]
```diff
@@ -43,6 +43,8 @@
 #define ELL_K_BITS  256ULL
 #define ELL_K_SIZE (ELL_K_BITS/8)
 
+#define BIKE1
```
Comment (andrewhop) [inline pq-crypto/bike/defs.h:46]: So I'm not attached either way, just a little curious why you switched to a define here rather than the makefile?

File: pq-crypto/sike/Makefile [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+# For now we ignore all of the errors below
+FLAGS_TO_REMOVE := -pedantic -Wall -Werror
+
+
+FINAL_CFLAGS := $(filter-out $(FLAGS_TO_REMOVE),$(CFLAGS))
+
+CFLAGS=$(FINAL_CFLAGS)
+CFLAGS+=-I../../
+
+SRCS=fp_generic.c P503.c sike_p503_kem.c fips202.c
```
Comment (andrewhop) [inline pq-crypto/sike/Makefile:26]: NP: wildcard here? That seems to be the standard for s2n.

File: tests/unit/s2n_sike_p503_kat_test.c [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ * Created by Bassham, Lawrence E (Fed) on 8/29/17.
+ * Copyright © 2017 Bassham, Lawrence E (Fed). All rights reserved.
+ */
+
+#include "s2n_test.h"
+#include "s2n_nist_kats.h"
+#include "pq-crypto/sike/sike_p503_kem.h"
+#include "pq-crypto/pq-random.h"
+
+#define RSP_FILE_NAME "kats/sike_p503.kat"
```
Comment (andrewhop) [inline tests/unit/s2n_sike_p503_kat_test.c:25]: NP: did you think about what a generic KAT test runner would look like? I was thinking if it would be cool to make a generic test that takes a file name, the 3 function pointers, and the sizes.

File: tests/unit/s2n_bike1_l1_kat_test.c [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        // Read the KAT values
+        EXPECT_SUCCESS(ReadHex(rsp_file, pk_answer, BIKE1_L1_PUBLIC_KEY_BYTES, "pk = "));
+        EXPECT_SUCCESS(ReadHex(rsp_file, sk_answer, BIKE1_L1_SECRET_KEY_BYTES, "sk = "));
+        EXPECT_SUCCESS(ReadHex(rsp_file, ct_answer, BIKE1_L1_CIPHERTEXT_BYTES, "ct = "));
+        EXPECT_SUCCESS(ReadHex(rsp_file, shared_secret_answer, BIKE1_L1_SHARED_SECRET_BYTES, "ss = "));
+
+        // Test the client and server got the same value
+        EXPECT_BYTEARRAY_EQUAL(client_shared_secret, server_shared_secret, BIKE1_L1_SHARED_SECRET_BYTES);
+
+        // Co\mpare the KAT values
```
Comment (andrewhop) [inline tests/unit/s2n_bike1_l1_kat_test.c:96]: NP: extra slash

File: tests/unit/s2n_sike_p503_kat_test.c [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+int main(int argc, char **argv, char **envp) {
+    BEGIN_TEST();
+
+    // Flip pq-random over to nist RNG to ensure KAT values match
+    EXPECT_SUCCESS(initialize_pq_crypto_generator(&randombytes));
```
Comment (andrewhop) [inline tests/unit/s2n_sike_p503_kat_test.c:31]: Did you look at pq-random.h? Do you think this is a good way to continue flipping between the RNGs? This is really a review of what I came up with and wondering what you think.

File: tests/unit/s2n_sike_p503_kat_test.c [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+#include "s2n_test.h"
+#include "s2n_nist_kats.h"
+#include "pq-crypto/sike/sike_p503_kem.h"
+#include "pq-crypto/pq-random.h"
+
+#define RSP_FILE_NAME "kats/sike_p503.kat"
+
+int main(int argc, char **argv, char **envp) {
+    BEGIN_TEST();
```
Comment (andrewhop) [inline tests/unit/s2n_sike_p503_kat_test.c:28]: Does SIKE work with the FIPS OpenSSL already? Should this skip the tests like BIKE? Or since we've accepted modifying the code should we apply your BIKE path that makes it work the correct FIPS OpenSSL API?

File: Makefile [hunk 7]
```diff
@@ -56,14 +56,7 @@ libs:
 
 .PHONY : bin
 bin: libs
```
Comment (andrewhop) [inline Makefile:58]: Was this change on purpose? Does bin not need any of these other directories to be built? It looks like the build is passing so I guess?

**Review by andrewhop — COMMENTED** [review]
Just some minor questions.

File: pq-crypto/bike/defs.h (hunk 1 above)
Comment (drucker-nir) [reply pq-crypto/bike/defs.h:46]: This made the CMAKE system a bit cleaner. No real effect on the Make system.
Because only BIKE1_L1 is supported then I prefer such definitions to be in the code and not in the (complicated) (c)make systems.

File: pq-crypto/sike/Makefile (hunk 2 above)
Comment (drucker-nir) [reply pq-crypto/sike/Makefile:26]: SIKE original implementation use a "weird" trick, it sources C files from within other C files, using a wildcard will cause double (and wrong) compilation of source files. I also think that wildcard is cleaner but I tried not to change anything in SIKE source code.

File: tests/unit/s2n_sike_p503_kat_test.c (hunk 3 above)
Comment (drucker-nir) [reply tests/unit/s2n_sike_p503_kat_test.c:25]: Yes, I already moved the utility functions to an external common header. I left the main flow as is (duplicated) because of two reasons:
  1) We only have two KAT tests (SIKE p503/BIKE1_L1)
  2) Passing function pointers, in some cases might complicate things and will need a special care.
In the trade off of having two small "main" tests or one that is a bit more complicated I chose the first option. If in the future we will decide to support additional levels/KEMs then making a generic test would be better. With that saying, feel free to change, I am not attached to either choice.

File: tests/unit/s2n_bike1_l1_kat_test.c (hunk 4 above)
Comment (drucker-nir) [reply tests/unit/s2n_bike1_l1_kat_test.c:96]: Thanks

File: tests/unit/s2n_sike_p503_kat_test.c (hunk 6 above)
Comment (drucker-nir) [reply tests/unit/s2n_sike_p503_kat_test.c:28]: To the best of my knowledge SIKE does not depends on OpenSSL at all (FIPS or not). 
I will apply BIKE patch to a future pull request.

File: Makefile (hunk 7 above)
Comment (drucker-nir) [reply Makefile:58]: Note that "bin" is calling "libs" ("bin: libs"), therefore all the internal libs are already compiled there. the code I removed was meaningless (compiling the same libraries again).
